### PR TITLE
Pipeline, Alexandra Eliseeva - 21FPL1

### PR DESCRIPTION
**Pull request by sashullel**

### CODE REVIEW

File: lab_6_pipeline/pipeline.py [hunk 1]
```diff
@@ -44,37 +88,58 @@ def __init__(self, lemma: str = "", pos: str = "", tags: str = ""):
...
+        self.tags = tags
 
 
 class ConlluToken:
     """
     Representation of the CONLL-U Token
     """
 
-    def __init__(self, text: str):
+    def __init__(self, text: str, parameters: MorphologicalTokenDTO, position: int):
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: you are not allowed to change API, discard

File: lab_6_pipeline/pipeline.py [hunk 2]
```diff
@@ -4,35 +4,79 @@
...
+        correct_range = [num for num in range(1, max(meta_files_idx) + 1)]
+        if meta_files_idx != correct_range != raw_files_idx:
+            raise InconsistentDatasetError('files are laid out in an inconsistent way')
 
     def _scan_dataset(self) -> None:
         """
         Register each dataset entry
         """
+        for file in self._path_to_raw_txt_data.glob('*_raw.txt'):
+            article = from_raw(file)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: assign by a key directly

File: lab_6_pipeline/pipeline.py [hunk 3]
```diff
@@ -4,35 +4,79 @@
...
         """
         Validates folder with assets
         """
+        if not self._path_to_raw_txt_data.exists():
+            raise FileNotFoundError('file does not exist')
+
+        if not self._path_to_raw_txt_data.is_dir():
+            raise NotADirectoryError('path does not lead to directory')
+
+        meta_files = [file for file in self._path_to_raw_txt_data.glob('*_meta.json')]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:55]: cast to list directly

File: lab_6_pipeline/pipeline.py [hunk 4]
```diff
@@ -144,16 +218,23 @@ def __init__(self, corpus_manager: CorpusManager):
...
         Returns the text representation as the list of ConlluSentence
         """
+        return [ConlluSentence(idx, sentence, []) for idx, sentence in enumerate(split_by_sentence(text))]
 
     def run(self) -> None:
         """
         Performs basic preprocessing and writes processed text to files
         """
+        for key, val in self._corpus.get_articles().items():
+            article = from_raw(val.get_raw_text_path(), val)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: it is already read by corpus manager (!)

File: lab_6_pipeline/pipeline.py [hunk 5]
```diff
@@ -86,6 +151,13 @@ def __init__(self, position: int, text: str, tokens: list[ConlluToken]):
         """
         Initializes ConlluSentence
         """
+        self._position = position
+        self._text = text
+        self._tokens = tokens
+
+    def _format_tokens(self, include_morphological_tags: bool) -> str:
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py:174]: do not see any usage, remove?

File: lab_6_pipeline/pipeline.py [hunk 6]
```diff
@@ -44,37 +88,58 @@ def __init__(self, lemma: str = "", pos: str = "", tags: str = ""):
...
+        return '\t'.join((self.position, self._text, self._morphological_parameters.lemma,
+                  self._morphological_parameters.pos, xpos, feats, head, deprel, deps, misc))
+
 
     def get_cleaned(self) -> str:
         """
         Returns lowercase original form of a token
         """
+        punc_to_remove = punctuation + '–—«»'
+        return self._text.translate(self._text.maketrans('', '', punc_to_remove)).lower()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: solve with regex, it is easier

**Review by demid5111 — CHANGES_REQUESTED** [review]
nice start, keep going

File: lab_6_pipeline/pipeline.py (hunk 1 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 6 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 2 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: right!

File: lab_6_pipeline/pipeline.py (hunk 1 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 2 above)
Comment (demid5111) [reply lab_6_pipeline/pipeline.py]: do not see it in code

File: lab_6_pipeline/pipeline.py (hunk 1 above)
Comment (demid5111) [reply lab_6_pipeline/pipeline.py]: do not see it in the code, have you committed and pushed?

File: lab_6_pipeline/pipeline.py [hunk 7]
```diff
@@ -144,16 +218,23 @@ def __init__(self, corpus_manager: CorpusManager):
         """
         Initializes MorphologicalAnalysisPipeline
         """
+        self._corpus = corpus_manager
 
     def _process(self, text: str) -> List[ConlluSentence]:
         """
         Returns the text representation as the list of ConlluSentence
         """
+        return [ConlluSentence(idx, sentence, []) for idx, sentence in enumerate(split_by_sentence(text))]
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: why do you leave tokens as empty list?

**Review by demid5111 — CHANGES_REQUESTED** [review]
no changes since last time, also I am afraid you have forgotten to push recent changes - I see answers in comments that are not reflected in code

File: lab_6_pipeline/pipeline.py (hunk 7 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: i couldn't understand the logic behind this method but now i do, fixed

File: lab_6_pipeline/pipeline.py [hunk 8]
```diff
@@ -1,38 +1,90 @@
 """
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: why is target score for pipeline still 0??

File: lab_6_pipeline/pipeline.py [hunk 9]
```diff
@@ -86,21 +169,36 @@ def __init__(self, position: int, text: str, tokens: list[ConlluToken]):
...
+        Formats tokens per newline
+        """
+        if include_morphological_tags:
+            return '\n'.join((token.get_conllu_text(include_morphological_tags) for token in self._tokens))
 
     def get_conllu_text(self, include_morphological_tags: bool) -> str:
         """
         Creates string representation of the sentence
         """
+        return (f'# sent_id = {self._position}\n'
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: use join with appropriate delimiter

File: lab_6_pipeline/pipeline.py [hunk 10]
```diff
@@ -144,16 +242,57 @@ def __init__(self, corpus_manager: CorpusManager):
...
         Initializes MorphologicalAnalysisPipeline
         """
+        self._corpus = corpus_manager
 
     def _process(self, text: str) -> List[ConlluSentence]:
         """
         Returns the text representation as the list of ConlluSentence
         """
+        conllu_sentences = []
+        mystem = Mystem()
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: define once per pipeline instance (in initializer?)

File: lab_6_pipeline/pipeline.py [hunk 11]
```diff
@@ -1,38 +1,90 @@
...
+
+        meta_files_idx = sorted(get_article_id_from_filepath(meta) for meta in meta_files)
+        raw_files_idx = sorted(get_article_id_from_filepath(raw) for raw in raw_files)
+        correct_range = [num for num in range(1, max(meta_files_idx) + 1)]
+        if meta_files_idx != correct_range != raw_files_idx:
+            raise InconsistentDatasetError('files are laid out in an inconsistent way')
+
+        for file in meta_files + raw_files:
+            with open(file, 'r', encoding='utf-8') as f:
+                if not f.read():
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: use .stat().st_size to understand whether file is empty or not. reading a file for it is overkill

File: lab_6_pipeline/pipeline.py [hunk 12]
```diff
@@ -144,16 +242,57 @@ def __init__(self, corpus_manager: CorpusManager):
...
+                      token['text'].strip() and token['text'] not in string.punctuation.replace('.', '')]
+            for token_idx, token_info in enumerate(tokens):
+                token_text = token_info['text']
+                print(token_text)
+                token = ConlluToken(token_text)
+                token.set_position(token_idx)
+
+                if token_info.get('analysis'):
+                    lemma = token_info['analysis'][0]['lex']
+                    pos = re.match('^[^,|=]*', token_info['analysis'][0]['gr']).group(0)
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: for pos processing you need to use mystem converter

**Review by demid5111 — CHANGES_REQUESTED** [review]
keep going!

File: lab_6_pipeline/pipeline.py (hunk 11 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: didn't use it because i thought we aren't allowed to use os module at all, done

**Review by demid5111 — CHANGES_REQUESTED** [review]
nothing new to review. hurry up, your PR is at risk of missing deadline.

have you pushed code to PR? I see no related changes that I expect to see according to your answers

File: lab_6_pipeline/pipeline.py (hunk 9 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: reworked

File: lab_6_pipeline/pipeline.py (hunk 10 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 12 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: done

File: lab_6_pipeline/pipeline.py (hunk 4 above)
Comment (sashullel) [reply lab_6_pipeline/pipeline.py]: got it!

File: lab_6_pipeline/data/mystem_tags_mapping.json [hunk 13]
```diff
@@ -0,0 +1,52 @@
+{
+  "POS": {
+    "A": "ADJ",
+    "ADV": "ADV",
+    "ADVPRO": "ADV",
+    "ANUM":	"ADJ",
```
Comment (demid5111) [inline lab_6_pipeline/data/mystem_tags_mapping.json]: layouting

File: lab_6_pipeline/data/opencorpora_tags_mapping.json [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    "ablt": "Ins",
+    "loct": "Ins",
+    "voct": "Voc",
+    "gen1": "Gen",
+    "gen2": "Par",
+    "loc1": "Loc",
+    "loc2": "Loc",
+    "acc2": "Acc"
+  }
+}
```
Comment (demid5111) [inline lab_6_pipeline/data/opencorpora_tags_mapping.json:53]: newline

File: lab_6_pipeline/pipeline.py [hunk 15]
```diff
@@ -4,35 +4,87 @@
...
+            raise InconsistentDatasetError('the number of raw and meta files does not match')
+
+        meta_files_idx = sorted(get_article_id_from_filepath(meta) for meta in meta_files)
+        raw_files_idx = sorted(get_article_id_from_filepath(raw) for raw in raw_files)
+        correct_range = [num for num in range(1, max(meta_files_idx) + 1)]
+        if meta_files_idx != correct_range != raw_files_idx:
+            raise InconsistentDatasetError('files are laid out in an inconsistent way')
+
+        for file in meta_files + raw_files:
+            if os.stat(file).st_size == 0:
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: use pathlib api

File: lab_6_pipeline/pipeline.py [hunk 16]
```diff
@@ -86,37 +169,78 @@ def __init__(self, position: int, text: str, tokens: list[ConlluToken]):
...
     """
+    def __init__(self, path: Path):
+        with open(path, 'r', encoding='utf-8') as f:
+            self._mapping_info = json.load(f)
 
     def convert_morphological_tags(self, tags: str) -> str:  # type: ignore
         """
         Converts the Mystem tags into the UD format
         """
+        if re.match('^[^,|=]*', tags).group(0) in \
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: rework to just focus on what you support

File: lab_6_pipeline/pipeline.py [hunk 17]
```diff
@@ -86,37 +169,78 @@ def __init__(self, position: int, text: str, tokens: list[ConlluToken]):
...
+                ('PART', 'PR', 'INTJ', 'CONJ', 'COM', 'ADVPRO', 'ADV'):
+            return ''
+
+        actual_tags = []
+        for tag in re.findall("[^()]+", tags):
+            actual_tag = re.split(r',|=', tag.split('|')[0])
+            actual_tags.extend(filter(None, actual_tag))
+
+        ud_tags = []
+        cats = {'NOUN': ['Animacy', 'Case', 'Gender', 'Number'],
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: defined in parent

File: lab_6_pipeline/pipeline.py [hunk 18]
```diff
@@ -144,16 +268,58 @@ def __init__(self, corpus_manager: CorpusManager):
...
 
     def _process(self, text: str) -> List[ConlluSentence]:
         """
         Returns the text representation as the list of ConlluSentence
         """
+        conllu_sentences = []
+        sentences = split_by_sentence(text)
+        for sent_idx, sentence in enumerate(sentences):
+            conllu_tokens = []
+            tokens = [token for token in self._mystem.analyze(re.sub(r'[^.\w\s]', '', sentence))
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: analyze first, clean then

File: lab_6_pipeline/pipeline.py [hunk 19]
```diff
@@ -144,16 +268,58 @@ def __init__(self, corpus_manager: CorpusManager):
...
+                token_text = token_info['text'].strip()
+                conllu_token = ConlluToken(token_text)
+                conllu_token.set_position(token_idx + 1)
+                if token_info.get('analysis'):
+                    lemma = token_info['analysis'][0]['lex']
+                    grammar_info = token_info['analysis'][0]['gr']
+                    pos = self._tag_converter.convert_pos(grammar_info)
+                    tags = self._tag_converter.convert_morphological_tags(grammar_info)
+                    parameters = MorphologicalTokenDTO(lemma=lemma.strip(), pos=pos, tags=tags)
+                else:
```
Comment (demid5111) [inline lab_6_pipeline/pipeline.py]: too nested

**Review by demid5111 — CHANGES_REQUESTED** [review]
comments are left

**Pull request closed without merging**